### PR TITLE
Add ProxyMeter to MeterFactoryBase which invokes real Meter if provided.

### PR DESCRIPTION
**Pull request by cijothomas**
One more step closer to configuration story for Meters. This allows us to write guidance for library authors to start emitting OT metrics, while OT Metric SDK specs are still evolving.

This updates the existing `MeterFactoryBase` to use a `ProxyMeter`, which would be "no-op" if no Meter is configured. When an actual `MeterFactory` is made available, by using `MeterFactoryBase.SetDefault`, the `ProxyMeter` will proxy calls to the real Meter obtained from the default `MeterFactory`.

**Libraries**
Take dependency on OT.API to emit metrics.
Obtain Meter like below:
```
var meter = MeterFactoryBase.Default.GetMeter("libraryname", "version").
// start emitting metrics using meter API. like createcounter, .add etc.  https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/metrics/api-user.md
```

**Applications**
Once MeterFactory is created with Exporter, Controller etc, call the following so that libraries will get the actual Meters.

// 1. create MeterFactory . Example:
```
var meterFactory = MeterFactory.Create(mb =>
                {
                mb.SetMetricProcessor(processor);
                mb.SetMetricExporter(promExporter);
                mb.SetMetricPushInterval(TimeSpan.FromSeconds(pushIntervalInSecs));
                });
```

2. Update the Default factory to the one created above.

```
MeterFactoryBase.SetDefault(meterFactory);
```

### CODE REVIEW

**Comment by cijothomas** [comment]
Note: I am not addressing renaming of Factory to Provider. (or rather create something new called MeterProvider, which internally uses the factory) This can be done after the same is addressed for Tracing.
https://github.com/open-telemetry/opentelemetry-dotnet/issues/576#issuecomment-618416278

File: src/OpenTelemetry.Api/Metrics/MeterFactoryBase.cs [hunk 1]
```diff
@@ -32,14 +37,50 @@ public static MeterFactoryBase Default
...
+        /// <param name="meterFactory">Instance of <see cref="MeterFactoryBase"/>.</param>
+        /// <remarks>
+        /// This method can only be called once. Calling it multiple times will throw an <see cref="System.InvalidOperationException"/>.
+        /// </remarks>
+        /// <exception cref="System.InvalidOperationException">Thrown when called multiple times.</exception>
+        public static void SetDefault(MeterFactoryBase meterFactory)
+        {
+            if (isInitialized)
+            {
+                throw new InvalidOperationException("Default factory is already set");
```
Comment (cijothomas) [inline src/OpenTelemetry.Api/Metrics/MeterFactoryBase.cs:51]: TracerFactory is doing this same thing - I think we can make it NOT throw, and log an error.
Will be following up on this aspect.

Comment (SergeyKanzhelev) [reply src/OpenTelemetry.Api/Metrics/MeterFactoryBase.cs:51]: I think it's worth discussing. The question is whether we want to hide this error or want to crash an app to indicate that something is very broken if it is being set twice. Worth having an issue for the discussion

**Comment by SergeyKanzhelev** [comment]
@cijothomas build is failing now =(

File: src/OpenTelemetry.Api/Metrics/MeterFactoryBase.cs (hunk 1 above)
Comment (cijothomas) [reply src/OpenTelemetry.Api/Metrics/MeterFactoryBase.cs:51]: https://github.com/open-telemetry/opentelemetry-dotnet/issues/638

**Pull request merged**